webapp/lib/main_ui.dart

```diff
@@ -45,9 +53,88 @@ class CodaUI {
       body.append(messageViewModel.viewElement);
     });
     messageCodingTable.append(body);
+
+    messageCodingTable.onChange.listen((event) {
+      var target = event.target;
+      if (target is! InputElement && target is! SelectElement) return;
+
+      TableRowElement row = getAncestors(CodeSelector.activeCodeSelector.viewElement).firstWhere((e) => e.classes.contains('message-row'));
+      DivElement inputGroup = getAncestors(CodeSelector.activeCodeSelector.viewElement).firstWhere((e) => e.classes.contains('input-group'));
+      String messageID = row.attributes['message-id'];
+      MessageViewModel message = messageMap[messageID];
+      String schemeID = inputGroup.attributes['scheme-id'];
+
+      if (target is InputElement) { // change on checkbox element
+        message.schemeCheckChanged(dataset, schemeID, target.checked);
+      }
+      if (target is SelectElement) { // change on dropdown select element
+        CodeSelector codeSelector = message.codeSelectors.singleWhere((codeSelector) => codeSelector.scheme.id == schemeID);
+        CodeSelector.activeCodeSelector = codeSelector;
+        message.schemeCodeChanged(dataset, schemeID, codeSelector.selectedOption);
+        codeSelector.hideWarning();
+        selectNextEmptyCodeSelector(messageID, schemeID);
+      }
+    });
+
+    window.onKeyDown.listen((event) {
+      if (event.key == 'Tab') {
+        TableRowElement row = getAncestors(CodeSelector.activeCodeSelector.viewElement).firstWhere((e) => e.classes.contains('message-row'));
+        String messageID = row.attributes['message-id'];
+        selectNextEmptyCodeSelector(messageID, CodeSelector.activeCodeSelector.scheme.id);
+        event.preventDefault();
+        event.stopPropagation();
+        return;
+      }
+
+      Map activeShortcuts = {};
+      CodeSelector.activeCodeSelector.scheme.codes.forEach((code) {
+        activeShortcuts[code['shortcut']] = code['valueID'];
+      });
+      activeShortcuts[' '] = CodeSelector.EMPTY_CODE_VALUE; // add space as shortcut for unassigning a code
+
+      if (activeShortcuts.keys.contains(event.key)) {
+        CodeSelector.activeCodeSelector.selectedOption = activeShortcuts[event.key];
+        CodeSelector.activeCodeSelector.hideWarning();
+        TableRowElement row = getAncestors(CodeSelector.activeCodeSelector.viewElement).firstWhere((e) => e.classes.contains('message-row'));
+        String messageId = row.attributes['message-id'];
+        messageMap[messageId].schemeCodeChanged(dataset, CodeSelector.activeCodeSelector.scheme.id, CodeSelector.activeCodeSelector.selectedOption);
+        selectNextEmptyCodeSelector(messageId, CodeSelector.activeCodeSelector.scheme.id);
+        event.preventDefault();
+        event.stopPropagation();
+        return;
+      }
+    });
+    CodeSelector.activeCodeSelector = messages[0].codeSelectors[0];
+  }
+
+  selectNextEmptyCodeSelector(String messageID, String schemeID) {
```

Comment:
I think splitting this into 2 methods, selectNextEmptyCodeSelectorVertical and selectNextEmptyCodeSelectorHorizontal would make sense and help decrease the level of if-else nesting

Reply:
That's a good idea, thanks!